### PR TITLE
Shop

### PR DESCRIPTION
**Pull request by kaashis**

### CODE REVIEW

File: src/main/java/com/zombiecastlerush/building/Castle.java [hunk 1]
```diff
@@ -12,28 +11,40 @@ public Castle() {
...
         drawBridge.setChallenge(new Puzzle("Draw-Bridge-Puzzle", "What is (2+2) X (2-2)?", "0"));
 
+        //add items to Rooms inventory
+        List<Item> shopItems = new ArrayList<>(Arrays.asList(
+                /* name    description                                                                                      price*/
+                new Item("Sword", "This is the sword of Destiny, made up of the Valyrian Steel", 100.0),
+                new Item("Helmet", "This is the ultimate shield which will be carried by Captain America in the distant future", 50.0),
+                new Item("Potion", "Drinking this potion will restore your health", 100.0)
+        ));
+        shop.getInventory().addItems(shopItems.toArray(new Item[0]));
```
Comment (chaynsaw) [inline src/main/java/com/zombiecastlerush/building/Castle.java]: Why do an `Arrays.asList` and a `.toArray()`? Why not just add the items in the inventory directly as varargs?

File: src/main/java/com/zombiecastlerush/building/Shop.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package com.zombiecastlerush.building;
+
+import com.zombiecastlerush.entity.Player;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Shop extends Room {
```
Comment (chaynsaw) [inline src/main/java/com/zombiecastlerush/building/Shop.java:9]: I realize that we never talked about what a Shop is, an interface with which you interact with in a room? Or a Room itself? I assumed former, it looks like you implemented latter, which seems workable now that I'm looking at it.

File: src/main/java/com/zombiecastlerush/util/Prompter.java [hunk 3]
```diff
@@ -61,10 +65,12 @@ static void advanceGame(Player player) throws JsonProcessingException {
                         System.out.println(player.displayStatus());
                     break;
                 case "pick-up":
-                    for (Item item : currentRoom.getInventory().getItems()) {
-                        if (item.getName().equalsIgnoreCase(userInputList.get(1))) {
-                            player.pickUp(item);
-                            break;
+                    if (!(currentRoom instanceof Shop)) {
```
Comment (chaynsaw) [inline src/main/java/com/zombiecastlerush/util/Prompter.java]: All the casting and the size of this switch case makes me uneasy, but I don't know how to fix it without a very involved refactoring... which is beyond the scope of this PR.

File: src/test/java/com/zombiecastlerush/building/ShopTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Test
+    public void testBuyWhenPlayerDoesNotHaveEnoughMoney() {
+        //can't buy when player does not have enough money
+        double preBuy_balance = p1.getAcctBalance();
+        s1.buy(p1, i1);
+        assertFalse(p1.getInventory().getItems().contains(i1));
+        assertTrue(s1.getInventory().getItems().contains(i1));
+        assertEquals(preBuy_balance, p1.getAcctBalance(), 0.001);
```
Comment (chaynsaw) [inline src/test/java/com/zombiecastlerush/building/ShopTest.java:32]: Would `assertEquals(0.001, p1.getAcctBalance())` not work here?

**Review by chaynsaw — COMMENTED** [review]
Minor issues, but I think it's generally there.

File: src/main/java/com/zombiecastlerush/building/Shop.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    public Shop(String name, String description) {
+        super(name, description);
+    }
+
+    public void buy(Player player, Item item) {
```
Comment (chaynsaw) [inline src/main/java/com/zombiecastlerush/building/Shop.java]: Maybe you should make `buy` into `sellItemToPlayer` and `sell` into `buyItemFromPlayer`. Also, perhaps if you can, eliminate the need for passing reference of player, can you possibly get the player reference from inheritance? If not, it's ok.

File: src/test/java/com/zombiecastlerush/building/ShopTest.java (hunk 4 above)
Comment (chaynsaw) [reply src/test/java/com/zombiecastlerush/building/ShopTest.java:32]: Disregard.

File: src/main/java/com/zombiecastlerush/building/Shop.java (hunk 5 above)
Comment (kaashis) [reply src/main/java/com/zombiecastlerush/building/Shop.java]: changed names accordingly

File: src/test/java/com/zombiecastlerush/building/ShopTest.java (hunk 4 above)
Comment (kaashis) [reply src/test/java/com/zombiecastlerush/building/ShopTest.java:32]: it works, but to handle deprecated msg from IDE for assertEqual(double,double)

**Review by chaynsaw — APPROVED** [review]

**Pull request merged**